lib/google/ads/google_ads/google_ads_client.rb

```diff
@@ -140,36 +131,73 @@ def service(name, version = default_api_version)
             end
             headers[:"login-customer-id"] = login_customer_id.to_s  # header values must be strings
           end
-          return class_to_return.new(
-            credentials: get_updater_proc(),
-            metadata: headers,
-            exception_transformer: ERROR_TRANSFORMER
-          )
+
+          if logger
+            logging_interceptor = Google::Ads::GoogleAds::LoggingInterceptor.new(logger)
+          end
+
+          if name.nil?
+            Factories::Services.new(
+              service_path: service_path,
+              logging_interceptor: logging_interceptor,
+              credentials: get_updater_proc,
+              metadata: headers,
+              exception_transformer: ERROR_TRANSFORMER
+            )
+          else
+            class_to_return = lookup_util.raw_service(name, version)
+            class_to_return = Class.new(class_to_return) do
+              unless service_path.nil? || service_path.empty?
+                const_set('SERVICE_ADDRESS', service_path.freeze)
+              end
+
+
+              const_set('GRPC_INTERCEPTORS', [logging_interceptor].compact)
+            end
+
+            class_to_return.new(
+              credentials: get_updater_proc,
+              metadata: headers,
+              exception_transformer: ERROR_TRANSFORMER
+            )
+          end
         end
 
         # Return a resource or common entity for the provided entity type. For
         # example, passing :Campaign will return an instantiated Campaign.
         #
         # Raises ArgumentError if no entity can be found for the provided type.
-        def resource(name, version = default_api_version)
-          lookup_util.resource(name, version)
+        def resource(name=nil, version=default_api_version)
+          if name.nil?
+            Google::Ads::GoogleAds::Factories::Resources
```

Comment:
Same for all these; you need to have one set of factories per version and look up the correct one for the given version.